allennlp/training/trainer.py

```diff
@@ -290,8 +297,130 @@ def on_end(
         **kwargs,
     ) -> None:
         if is_primary:
-            assert self._tensorboard is not None
-            self._tensorboard.close()
+            assert self._log_writer is not None
+            self._log_writer.close()
+
+
+@TrainerCallback.register("console-logger")
+class ConsoleLoggerCallback(TrainerCallback):
+    def __init__(
+        self,
+        serialization_dir: Optional[str] = None,
+        should_log_inputs: bool = False,
+    ) -> None:
+        super().__init__(serialization_dir=serialization_dir)
+        self._should_log_inputs = should_log_inputs
+
+    def on_batch(
+        self,
+        trainer: "GradientDescentTrainer",
+        batch_inputs: List[List[TensorDict]],
+        batch_outputs: List[Dict[str, Any]],
+        batch_metrics: Dict[str, Any],
+        epoch: int,
+        batch_number: int,
+        is_training: bool,
+        is_primary: bool = True,
+        batch_grad_norm: Optional[float] = None,
+        **kwargs,
+    ) -> None:
+        # In the distributed case we need to call this from every worker, since every
+        # worker reports its own memory usage.
+        cpu_memory_usage = common_util.peak_cpu_memory()
+        gpu_memory_usage = common_util.peak_gpu_memory()
+
+        if not is_primary:
+            return None
+
+        # We only want to do this for the first batch (TODO: only first epoch too?)
+        if batch_number == 1:
+
+            # Log memory usage
+            cpu_memory_usage_total = 0.0
+            for worker, mem_bytes in cpu_memory_usage.items():
+                memory = mem_bytes / (1024 * 1024)
+                logger.info(f"memory_usage/worker_{worker}_cpu - {memory}")
+                cpu_memory_usage_total += memory
+            logger.info(f"memory_usage/cpu - {cpu_memory_usage_total}")
+            for gpu, mem_bytes in gpu_memory_usage.items():
+                memory = mem_bytes / (1024 * 1024)
+                logger.info(f"memory_usage/gpu_{gpu} - {memory}")
+
+            # Log batch inputs
+            if self._should_log_inputs:
+                logger.info("Batch inputs")
+                for b, batch in enumerate(batch_inputs):
+                    self._log_fields(batch, log_prefix="batch_input")  # type: ignore
+
+    def _log_fields(self, fields: Dict, log_prefix: str = ""):
+        for key, val in fields.items():
+            key = log_prefix + "/" + key
+            if isinstance(val, dict):
+                self._log_fields(val, key)
+            elif isinstance(val, torch.Tensor):
+                logger.info(f"\"{key}\" (Shape : {' x '.join([str(x) for x in val.shape])})")
+                torch.set_printoptions(threshold=2)
+                logger.info(f"{val}")
+                torch.set_printoptions(threshold=1000)
```

Comment:
This looks a little weird in the logs because of how the representation of the tensor is split over multiple lines:

![image](https://user-images.githubusercontent.com/8812459/108921367-13824180-75eb-11eb-9aaa-ff3ba472f6fa.png)

Maybe something like this:

```suggestion
                torch.set_printoptions(threshold=2)
                logger.info("%s (Shape: %s)\n%s", key, " x ".join([str(x) for x in val.shape]), val)
                torch.set_printoptions(threshold=1000)
```

Which gives you:

![image](https://user-images.githubusercontent.com/8812459/108921822-6e1b9d80-75eb-11eb-9133-0de8d3ca5bd2.png)

Feel free to implement this differently (see my comment about my preference for avoiding f-strings below).